src/ballet/block/fd_microblock.h

```diff
@@ -0,0 +1,180 @@
+
+/* fd_microblock_new formats an unused memory region for storing an
+   `fd_microblock_t`. `shmem` is a non-NULL pointer to this region in the
+   local address space with the required footprint and alignment.
+   `txn_max_cnt` is the number of txn descriptors that can be stored
+   in the `fd_microblock_t`.  The `fd_microblock_hdr_t` is initialized to zero.
+   Returns `shmem` on success and NULL on failure (logs details).
+   Reasons for failure include an obviously bad `shmem` (alignment)
+   or bad `txn_max_cnt`. */
+
+void *
+fd_microblock_new( void * shmem,
+                   ulong  txn_max_cnt );
+
+/* `fd_microblock_join` joins the caller to the microblock.
+   `shblock` points to the first byte of the memory region backing the
+   `fd_microblock_t` in the caller's address space.
+   Returns `shblock` on success and NULL on failure (logs details).
+   Reasons for failure include those of `fd_microblock_new` or that the
+   memory region was not correctly initialized.
+   Every successful join should have a matching leave. The lifetime of
+   the join is until the matching leave or thread group is terminated. */
+
+fd_microblock_t *
+fd_microblock_join( void * shblock );
+
+/* `fd_microblock_leave` leaves a current local join.  Returns a pointer to
+   the underlying shared memory region on success.  Reasons for failure
+   include `block` is NULL. */
+
+void *
+fd_microblock_leave( fd_microblock_t * block );
+
+/* `fd_microblock_delete` unformats a memory region storing an `fd_microblock_t`.
+   Assumes nobody is joined to the region.  Returns a pointer to the
+   underlying shared memory region or NULL if used obviously in error
+   (e.g. `shblock` is obviously not an `fd_microblock_t` ... logs details).
+   The ownership of the memory region is transferred to the caller. */
+
+void *
+fd_microblock_delete( void * shblock );
+
+/* fd_microblock_deserialize: Deserializes a microblock from `buf`.
+   Only accesses up to `buf_sz` bytes beyond `buf`.
+
+   Returns 1 on success and 0 on failure.
+   Reasons for failure include invalid data or unexpected EOF.
+
+   If the returned ptr is non-NULL, it is guaranteed to be
+     1) greater than the provided `buf` ptr
+     2) less or equal than `(uchar *)buf + buf_sz` */
+int
+fd_microblock_deserialize( fd_microblock_t * block,
+                           void **           buf,
+                           ulong *           buf_sz,
+                           fd_txn_parse_counters_t * counters_opt );
+
+/* fd_microblock_mixin: Calculates the PoH mixin hash.
+
+   Computes the root of a 32-byte binary Merkle tree of a vector with
+   each element containing the first Ed25519 signature of each txn.
+
+   For microblocks that do not contain transactions, a single append
+   instead of a mixin should be applied to PoH.
+
+   U.B. if this microblock contains a txn with zero signatures (illegal txn).
+   U.B. if this microblock's `hdr.txn_cnt` is zero. */
```

Comment:
Would it make sense for this code to handle the 0 transactions case? Is there a reason it can't?